### PR TITLE
adding k shortest paths implementation

### PR DESCRIPTION
**Pull request by uditagar**
This is an implementation for finding k shortest paths (paths can contain cycles) from start to end node.

### CODE REVIEW

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+using OutEdgeRangeFn = SSSP::OutEdgeRangeFn;
+using TileRangeFn = SSSP::TileRangeFn;
+
+using PSchunk = katana::PerSocketChunkFIFO<kChunkSize>;
+using OBIM = katana::OrderedByIntegerMetric<UpdateRequestIndexer, PSchunk>;
+using OBIM_Barrier = katana::OrderedByIntegerMetric<
+    UpdateRequestIndexer, PSchunk>::with_barrier<true>::type;
+
+bool
+CheckIfReachable(Graph* graph, const GNode& source) {
```
Comment (gurbinder533) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: `CheckReachability`?

Comment (uditagar) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: will do

File: lonestar/analytics/cpu/k-shortest-paths/README.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+2. Run `cd <BUILD>/lonestar/analytics/cpu/k-shortest-paths; make -j`
+
+RUN
+--------------------------------------------------------------------------------
+
+The following are a few example command lines.
+
+-`$ ./k-shortest-paths-cpu <path-to-graph> --algo=deltaStep --delta=13 --edgePropertyName=value --numPaths=10 --startNode=1 --reportNode=100 -t 40 
+-`$ ./k-shortest-paths-cpu <path-to-graph> --algo=deltaTile --delta=13 --edgePropertyName=value --numPaths=10 --startNode=1 --reportNode=100 -t 40`
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/README.md]: Why two identical lines?

File: lonestar/analytics/cpu/k-shortest-paths/README.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+PERFORMANCE  
+--------------------------------------------------------------------------------
+
+* deltaStep/deltaTile algorithms typically performs the best on high diameter
+  graphs, such as road networks. Its performance is sensitive to the *delta* parameter, which is
+  provided as a power-of-2 at the commandline. *delta* parameter should be tuned
+  for every input graph
+* topo/topoTile algorithms typically perform the best on low diameter graphs, such
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/README.md]: Drop this point

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+
+struct NodeCount {
+  using ArrowType = arrow::CTypeTraits<uint32_t>::ArrowType;
+  using ViewType = katana::PODPropertyView<std::atomic<uint32_t>>;
+};
+
+struct NodeMax {
+  using ArrowType = arrow::CTypeTraits<uint32_t>::ArrowType;
+  using ViewType = katana::PODPropertyView<std::atomic<uint32_t>>;
+};
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp:89]: Replace with:
```
using NodeCount = UInt32Property;
using NodeMax = UInt32Property;
```

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+
+struct EdgeWeight : public katana::PODProperty<uint32_t> {};
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: Why not?
`using EdgeWeight = UInt32Property;`

File: lonestar/analytics/cpu/k-shortest-paths/README.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  as social networks and RMAT graphs
+* All algorithms rely on CHUNK_SIZE for load balancing,
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/README.md]: Change `CHUNK_SIZE` to `kChunkSize`

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+bool
+CheckReachability(Graph* graph, const GNode& source) {
+  katana::InsertBag<GNode> current_bag;
+  katana::InsertBag<GNode> next_bag;
+
+  current_bag.push(source);
+  graph->GetData<NodeCount>(source) = 1;
+
+  while (true) {
+    if (current_bag.begin() == current_bag.end())
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: `while (current_bag.begin() != current_bag.end())`

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+      break;
+
+    katana::do_all(
+        katana::iterate(current_bag),
+        [&](GNode n) {
+          for (auto edge : graph->edges(n)) {
+            auto dest = *(graph->GetEdgeDest(edge));
+            if (graph->GetData<NodeCount>(dest) == 0) {
+              graph->GetData<NodeCount>(dest) = 1;
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp:203]: This is just BFS right? Maybe add an option for asynchronous `for_each`?

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+    current_bag.clear();
+    std::swap(current_bag, next_bag);
+  }
+
+  if (graph->GetData<NodeCount>(reportNode) == 0) {
+    return false;
+  }
+
+  katana::do_all(katana::iterate(*graph), [&graph](GNode n) {
+    graph->GetData<NodeCount>(n) = 0;
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp:219]: Are you sure you don't want to do this if you're returning false?

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+  katana::GAccumulator<size_t> bad_work;
+  //! [reducible for self-defined stats]
+  katana::GAccumulator<size_t> wl_empty_work;
+
+  graph->GetData<NodeCount>(source) = 1;
+
+  katana::InsertBag<Item> init_bag;
+
+  Path* path = new Path();
+  path->last = NULL;
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: `nullptr`

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+          path->last = item.path;
+          path_pointers->push(path);
+
+          if (ddata_count < numPaths) {
+            katana::atomicAdd<uint32_t>(ddata_count, (uint32_t)1);
+            katana::atomicMax<uint32_t>(ddata_max, new_dist);
+          }
+
+          if (dst == reportNode)
+            bag->push(std::make_pair(new_dist, path));
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: `if (...) {}`

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+  return true;
+}
+
+//delta stepping implementation for finding a shortest path from source to report node
+template <typename Item, typename OBIMTy, typename PushWrap, typename EdgeRange>
+void
+DeltaStepAlgo(
+    Graph* graph, const GNode& source, const PushWrap& pushWrap,
+    const EdgeRange& edgeRange,
+    katana::InsertBag<std::pair<uint32_t, Path*>>* bag,
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: A more descriptive name please. Like report_node_bag?

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+            katana::atomicAdd<uint32_t>(ddata_count, (uint32_t)1);
+            katana::atomicMax<uint32_t>(ddata_max, new_dist);
+          }
+
+          if (dst == reportNode)
+            bag->push(std::make_pair(new_dist, path));
+
+          if ((graph->GetData<NodeCount>(reportNode) < numPaths) ||
+              ((graph->GetData<NodeCount>(reportNode) >= numPaths) &&
+               (graph->GetData<NodeMax>(reportNode) > new_dist))) {
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: A more clear way to write this:
```
bool terminate = ...;
if (!terminate) {
  ...
  pushWrap(..);
}
```
More comments would help too.

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+            const Path* const_path = path;
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp:285]: Why is this required?

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp (hunk 13 above)
Comment (roshandathathri) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: There's another way to terminate. `for_each` has support for a parallel break/exit. 
So a more clear way to write this is:
```
for_each(..., [&](const Item& item, auto& ctx) {
for (auto ii : edgeRange(item)) {
  GNode dst = *(graph->GetEdgeDest(ii));
  ...
  if (dst == reportNode) {
    ... 
    bool terminate = ...;
    if (terminate) {
      ctx.breakLoop();
      return;
    }
  }
  ...
  pushWrap(...);
}
}, katana::parallel_break(), ...);
```

However, I'm not sure if this would be more performant than the non-break option above. So please benchmark both before you choose one.

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 15]
```diff
@@ -0,0 +1,345 @@
...
+    katana::gInfo("Using delta-step of ", (1 << stepShift), "\n");
+    KATANA_LOG_WARN(
+        "Performance varies considerably due to delta parameter.\n");
+    KATANA_LOG_WARN("Do not expect the default to be good for your graph.\n");
+  }
+
+  katana::do_all(katana::iterate(graph), [&graph](GNode n) {
+    graph.GetData<NodeMax>(n) = 0;
+    graph.GetData<NodeCount>(n) = 0;
+  });
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: Please push this to a function called `Initialize*()`

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 16]
```diff
@@ -0,0 +1,345 @@
...
+
+    default:
+      std::abort();
+    }
+  }
+
+  execTime.stop();
+
+  if (reachable) {
+    std::multimap<uint32_t, Path*> paths_map;
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp:423]: Is this an ordered map?

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 17]
```diff
@@ -0,0 +1,345 @@
...
+    for (uint32_t iter = 0; iter < num; iter++) {
+      katana::gPrint(" ", report);
+      GNode parent = report;
+
+      const Path* path = it_report->second;
+      it_report++;
+
+      while (path->last != NULL) {
+        parent = path->parent;
+        katana::gPrint(" ", parent);
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: nit: I think it's better to print a path from source to report node. Please write a recursive function to print the path in such a way.

Also please print the distance of the path.

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 18]
```diff
@@ -0,0 +1,345 @@
...
+        path = path->last;
+      }
+      katana::gPrint("\n");
+    }
+
+    katana::do_all(
+        katana::iterate(path_pointers), [&](Path* p) { delete (p); });
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: I had originally typed this: "Drop `path_pointers` and change all `Path*` to `std:;shared_ptr<Path>`. We should avoid using plain/C-style pointers". However, with the use of custom allocation, `std::shared_ptr` cannot be used. So, this is fine for now.

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 19]
```diff
@@ -0,0 +1,345 @@
...
+          auto& ddata_max = graph->GetData<NodeMax>(dst);
+
+          Distance ew = graph->GetEdgeData<EdgeWeight>(ii);
+          const Distance new_dist = item.distance + ew;
+
+          if ((ddata_count >= numPaths) && (ddata_max <= new_dist))
+            continue;
+
+          Path* path;
+          path = new Path();
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: Concurrent allocation using `new` would be bad in performance. Use our custom allocators instead. Try something like:
```
class PathAlloc {
public:
  Path* NewPath() {
     Path* path = allocator_.allocate(1);
     allocator_.construct(path, Path());
     return path;
  }

  void DeletePath(Path* path) {
     allocator_.destroy(path);
     allocator_.deallocate(path, 1);
  }

private:
  katana::FixedSizeAllocator<Path> alloc_;
};
```

Then instead of calling `new Path()` and `delete (path)`, call `path_alloc.NewPath()` and `path_alloc.DeletePath(path)`. Declare `PathAlloc path_alloc;` before `for_each`.

File: lonestar/analytics/cpu/k-shortest-paths/README.md (hunk 2 above)
Comment (uditagar) [reply lonestar/analytics/cpu/k-shortest-paths/README.md]: not identical. It just shows two different algo types

Comment (roshandathathri) [reply lonestar/analytics/cpu/k-shortest-paths/README.md]: Sorry for missing that!

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp (hunk 4 above)
Comment (uditagar) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp:89]: sorry, I don't understand this one. If you can further explain, that would be good. Thank you

Comment (uditagar) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp:89]: this will probably not allow atomic adds, I think

Comment (roshandathathri) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp:89]: Yes, I had missed that. This is fine as is for now.

But I think we should add `std::atomic` variants for property views in Properties.h in the long run. @ddn0

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp (hunk 9 above)
Comment (uditagar) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp:219]: yes, because there is nothing more to do if there is no path from source to report node and thus the code terminates, when false is returned,

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp (hunk 13 above)
Comment (uditagar) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: I think my intention was not to terminate but to check whether I want to put this new extended path to the worklist. And I guess your second suggestion would probably exit the for loop, right? I might be wrong here too. thanks

Comment (roshandathathri) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: Yes, both yield the same results.

Comment (uditagar) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: but I am not trying to check for termination. I do admit that I should have written it in a better way, like your first suggestion.

Comment (roshandathathri) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: Why not? It's safe to terminate as soon as you have the required number of paths for the report node.

Comment (uditagar) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: the issue is that there is still a chance that I can a get a shorter path later on. So I need to run it until all items in the worklist are processed.

Comment (roshandathathri) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: Makes sense. Yes, it's not a termination condition. I get it now.

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp (hunk 16 above)
Comment (uditagar) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp:423]: yes, it's ordered

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp (hunk 19 above)
Comment (uditagar) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: I am getting runtime errors with this approach. It will be good if you can give me some suggestion. thanks

Comment (roshandathathri) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: What error are you getting?

It definitely supports concurrent allocations. Check delaunay triangulation for an example usage.

Comment (uditagar) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: Thanks man. I will go through it then.

Comment (uditagar) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: I get this segmentation fault: `Thread 1 "k-shortest-path" received signal SIGSEGV, Segmentation fault.
katana::PerBackend::getRemote (this=0x55555592a4c0 <katana::getPTSBackend()::b>, thread=0, 
    offset=3968) at /home/udit/udit_katana_public_Jan5/libgalois/src/PerThreadStorage.cpp:155
155	  char* rbase = heads[thread].load(std::memory_order_relaxed);`

Comment (roshandathathri) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: I can help debug this. Can you please push your code?

Comment (uditagar) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: I just pushed it. Thank you.

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 20]
```diff
@@ -0,0 +1,457 @@
...
+const char* const ALGO_NAMES_BFS[] = {"asyncBFS", "syncBFS"};
+
+static cll::opt<AlgoBFS> algoBFS(
+    "algoBFS", cll::desc("Choose an algorithm:"),
+    cll::values(clEnumVal(asyncBFS, "asyncBFS"), clEnumVal(syncBFS, "syncBFS")),
+    cll::init(syncBFS));
+
+struct Path {
+  uint32_t parent;
+  const Path* last;
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]:
```cpp
const Path* last{nullptr};
```

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 21]
```diff
@@ -0,0 +1,457 @@
...
+  //! [reducible for self-defined stats]
+  katana::GAccumulator<size_t> wl_empty_work;
+
+  graph->GetData<NodeCount>(source) = 1;
+
+  katana::InsertBag<Item> init_bag;
+
+  //Path* path = new Path();
+  Path* path;
+  path_alloc.NewPath();
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]:
```cpp
Path* path = path_alloc.NewPath();
```

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 22]
```diff
@@ -0,0 +1,457 @@
...
+  path->last = nullptr;
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp:243]: Drop line. It is not required with the default initialization I mentioned above.

However, what is the value of parent for this path?

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 23]
```diff
@@ -0,0 +1,457 @@
...
+
+          Distance ew = graph->GetEdgeData<EdgeWeight>(ii);
+          const Distance new_dist = item.distance + ew;
+
+          if ((ddata_count >= numPaths) && (ddata_max <= new_dist))
+            continue;
+
+          Path* path;
+          //path = new Path();
+          path_alloc.NewPath();
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: `Path* path = path_alloc.NewPath();`

**Review by roshandathathri — COMMENTED** [review]
Try if these changes fixes the issue.

**Comment by uditagar** [comment]
> Try if these changes fixes the issue.

Thanks, I will check them now.

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp (hunk 22 above)
Comment (uditagar) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp:243]: this path pointer corresponds to the source node, hence no parent to keep track of.

**Comment by uditagar** [comment]
Thank you for the comments. It seems I made some stupid mistake there. It seems to be working now.

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp (hunk 22 above)
Comment (roshandathathri) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp:243]: It's not good to leave the parent uninitialized. Maybe just assign source as the parent?

**Comment by roshandathathri** [comment]
No worries. `NewPath()` returns a pointer just like `new Path()`

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp (hunk 22 above)
Comment (uditagar) [reply lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp:243]: ok, I will do that

**Comment by uditagar** [comment]
I have incorporated all the comments now. Thank you

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 24]
```diff
@@ -0,0 +1,452 @@
...
+    cll::desc("Number of paths to compute from source to report node (default "
+              "value 1)"),
+    cll::init(1));
+enum AlgoSSSP { deltaTile = 0, deltaStep, deltaStepBarrier };
+
+const char* const ALGO_NAMES_SSSP[] = {
+    "deltaTile", "deltaStep", "deltaStepBarrier"};
+
+static cll::opt<AlgoSSSP> algoSSSP(
+    "algoSSSP", cll::desc("Choose an algorithm:"),
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: -> "Choose an algorithm for SSSP"

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 25]
```diff
@@ -0,0 +1,452 @@
...
+    cll::values(
+        clEnumVal(deltaTile, "deltaTile"), clEnumVal(deltaStep, "deltaStep"),
+        clEnumVal(deltaStepBarrier, "deltaStepBarrier")),
+    cll::init(deltaTile));
+
+enum AlgoBFS { asyncBFS = 0, syncBFS };
+
+static cll::opt<AlgoBFS> algoBFS(
+    "algoBFS", cll::desc("Choose an algorithm:"),
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: -> "Choose an algorithm for reachability"

BFS and SSSP is confusing. Use "reachability" instead of bfs.

File: lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp [hunk 26]
```diff
@@ -0,0 +1,452 @@
...
+    cll::values(clEnumVal(asyncBFS, "asyncBFS"), clEnumVal(syncBFS, "syncBFS")),
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp]: Drop BFS in the string names

**Review by roshandathathri — APPROVED** [review]
Thanks, Udit. LGTM
